### ISSUE TITLE
Implement some form of escape analysis to allocate values on the stack

### ISSUE THREAD

**Issue by yorickpeterse** [inko-lang/inko#776]
### Description

Types defined using the `class` keyword are heap allocated. While #750 talks about giving users the option to explicitly opt-in to stack allocating data, I feel we also need a mechanism to determine if we can do this automatically. That is, we need some form of escape analysis and stack allocate values if we determine this is safe to do so.

An area of the standard library that can really benefit from this is the `std.iter` module. Iterators are used heavily and incur a heap allocation. When used in tight loops, it's not uncommon for most time to be dominated by the memory management necessary for these iterators.

The basic idea is simple: if an allocated value `V` doesn't escape the scope `S` it's allocated into (meaning `lifetime(V) <= lifetime(S)`), we can allocate it onto the stack instead.

The implementation however is a bit more tricky. For example, we can only stack allocate values if they are allocated directly in their scope and not through a method call. This means that good inlining is crucial for escape analysis to have a meaningful impact. In addition, the implementation of escape analysis itself may prove a bit of a challenge.

Note that it's perfectly fine for _borrows_ of stack allocated data to outlive the stack value, as dropping the stack value would still incur a runtime borrow check and produce a panic in such cases. In other words, such cases don't inhibit the ability to stack allocate.

**Comment by yorickpeterse** [comment]
A related and older idea is to bump allocate small objects using a custom allocator (https://github.com/inko-lang/inko/issues/542). The idea here was that by using our own allocator we can avoid the cost of going through `malloc` & friends. While this may not perform better compared to using e.g. jemalloc, glibc's allocator in particular is really not great when handling large amounts of allocations rapidly. This means it might be worth using for small objects (e.g. up to 64 bytes) that we can't stack allocate.

**Comment by yorickpeterse** [comment]
As for what constitutes as an escape: a conservative approach would be to just consider any move (except for a drop) an escape. This does mean that moving stack value S₁ into stack value S₂ results in S₁ being heap allocated, but it might be good enough as a start.

**Comment by yorickpeterse** [comment]
The way we generate MIR poses a challenge, such that we can't _just_ look at `move` instructions. That is, code like this:

```inko
let a = User(name: 'Alice')
```

Results in something like this:

```
R₁ = allocate User
R₁.name = 'Alice'
R₂ = move R₁
```

This means that if we look at just `move` instructions, it might not be accurate/sufficient enough to determine if we can stack allocate `User`.

**Comment by yorickpeterse** [comment]
Assuming the moves are all in the same scopes, it doesn't _really_ matter how many moves there are, since we can copy the data if needed. Whether copying actually happens depends largely on how LLVM decides to compile/optimize the code.

Things do get tricky if we borrow in between such moves, because the move would invalidate the borrow.

**Comment by yorickpeterse** [comment]
Given the following instructions:

```
R₁ = allocate User(...)
R₂ = move R₁
R₃ = move R₂
...
```

R₁ should be stack allocated, but the decision to do that depends on R₂, R₃ and others _not_ escaping the scope R₁ is defined in. So given a move chain of `A -> B -> C -> ...`, if all of those happen in the same scope or a nested scope, we can stack allocate `A`.

For this to work, for every register R that's assigned the result of an `allocate`, we need to track the registers it's moved to (recursively), and record the scopes those moves take place in.

**Comment by yorickpeterse** [comment]
For https://github.com/inko-lang/inko/issues/677 I have a WIP parser that uses the commonly used `strftime()`/`strptime()` format. This parser allocates many `Option[Int]` instances, where the `Int` is a byte.

Stack allocating these values would likely greatly improve performance. However, I'm not sure it's enough to _just_ rely on escape analysis. That is, for that to work the allocations need to be inlined, and given the complexity of the parser that's not a guarantee.

An alternative might be to not rely on escape analysis at all, but instead look at the type definition. That is, if the type size is e.g. less than or equal to 64 bytes, and all its sub values are trivial to copy, we just always stack allocate it and copy it upon borrowing. This does mean we need some sort of `mut T` where `T` _is_ on the stack but the borrow is an actual pointer, such that mutable methods can mutate the data in-place.

**Comment by yorickpeterse** [comment]
I've spent some time building a bump allocator for small objects based on Immix, as found on the `bump-allocator` branch. While a naive bump allocator can improve performance in allocation heavy applications, the moment you throw in reuse/recycling of memory the improvements are reduced greatly. This is because reusing memory requires additional bookkeeping, which in turn imposes a cost on both the allocation and deallocation path.

For the work in progress `DateTime.parse` implementation the bump allocator reduces the time by about 30%, but for https://github.com/jinyus/related_post_gen there's little to no difference. A key issue there seems to be that it simply requires a lot of objects and thus blocks, and thus most time is spent allocating those blocks. Increasing the block size helps a little bit, but it's still outperformed by using jemalloc (310 ms vs 325-330 ms for my bump allocator).

I'm now wondering if we perhaps could take a simpler approach. That is, instead of relying on escape analysis we instead promote types to value types based on their _definition_ rather than their usage. For example, types such as `Option[Int]` could easily be turned into value types and passed around using copies, removing the need for runtime memory allocation entirely.

The downside is that "value type promotion" only works for simple types such as `Option[Int]` but not for more complex types such as `Array[Int]`, while bump allocation _could_ help with more complex types. Stack allocation would also be helpful for this, but is likely to be incorrect (as in, not promoting values to the stack) more often than desired.

**Comment by yorickpeterse** [comment]
Another issue with value type promotion: if we just blindly promote a type to a value type (assuming it stores other value types), then we might break circular types (i.e. a linked list).

**Comment by yorickpeterse** [comment]
For the bump allocation scheme, I think we can make the following changes:

1. When finding the next hole to allocate into, simply just find the next line and that's it, instead of trying to find a hole consisting of multiple lines. I think the total cost might be roughly the same, but at least the hole finding logic will be simpler.
2. Finding reusable blocks whenever we exhaust the last block is too expensive, as we can end up with the pattern "find no reusable blocks -> allocate new block -> exhaust block -> find no reusable blocks -> repeat". A better approach may be to only do this every N block allocations
3. The mechanism for finding reusable lines using atomics is way too expensive, with a worst-case of 254 atomic loads. This needs to go, but I'm not sure yet how.

**Comment by yorickpeterse** [comment]
One option is to flip things around: values go on the _stack_ by default, and a hypothetical `box expr` expression puts it on the heap (e.g. `box [10, 20, 30]` or `box SomeThing.new(...)`). `class` instances still have headers and allow for dynamic dispatch, and are still subject to single ownership. Casting `class` values to a trait isn't possible and requires explicit boxing first.

To make that safe, we'd have to check the borrow count upon moving a stack value such that we don't invalidate any. In theory we can elide such checks if we can statically determine no borrows are present at the time of the move, but this won't be able to remove all such checks. Moving `box T` values around incurs no borrow check. Stack values are specialized over their _size_ not the type, such that 15 different types of the same size being passed to a generic type results in only one copy of the type.

Implementation wise this is much simpler compared to escape analysis and value type promotion, and we drastically reduce the pressure on the system allocator. This in turn is likely to improve performance significantly.

The downsides are:

- More runtime borrow checks when moving stack data around
- I'm not sure we can strictly guarantee our borrow checks always occur before a move, because LLVM optimizations might reorder things. We might have to insert some sort of compiler fence to prevent that from happening
- A more complicated type system, as users now have to deal with the differences between `User`, `box User`, etc
- Recursive types now explicitly require boxing, e.g. a linked list `Node` must be defined as `class Node[T] { let @value: T, let @next: Option[box Node[T]] }`

**Comment by yorickpeterse** [comment]
The above idea won't work well for resizable types such as arrays: if you have an array of N stack values and you create a borrow to a value, then resizing the array will invalidate that borrow. To make that safe as well, we'd have to check the borrow count for every value, which will be terrible performance wise.

**Comment by yorickpeterse** [comment]
Enums could be promoted to stack based value types _if_ the following conditions are met:

- The values they wrap (if any) are also stack types
- The type is not a recursive type

The reason this would be safe is that enums themselves can't be modified in-place, i.e. you can't change the constructor tag or the constructor's arguments. You _can_ modify those in-place technically, but for value types that would result in a copy being modified.

Essentially the idea here is that in addition to introducing an `inline` keyword to flag a type as an immutable value type, we'd infer this for `enum` types, a bit like a hypothetical `class enum inline? Foo { ... }` definition where `inline?` roughly translates to "inline whenever possible, and fall back to a heap value otherwise".

**Comment by yorickpeterse** [comment]
I started work on supporting stack/inline types in https://github.com/inko-lang/inko/pull/778. For my WIP implementation of `DateTime.parse` (https://github.com/inko-lang/inko/issues/677), the use of such `inline` types improves performance by 45-50%, as it no longer needs to allocate `Option[Int]` values all over the place.

**Comment by yorickpeterse** [comment]
https://github.com/inko-lang/inko/pull/778 is merged, so the bulk of the work is done. I do think it's still worth implementing stack allocations _on top_ of this. This way if a heap allocated type is used but we can trivially determine it should go on the stack, we can do just that automatically.

**Comment by yorickpeterse** [comment]
Escape analysis may be tricky to implement in an effective way: when encountering a dynamic dispatch call site we'd have to conservatively assume any arguments passed (including the receiver) escape the call, as we have no way of knowing which method is actually called at runtime. Since we use dynamic dispatch for most generics, this means that escape analysis might end up not being that useful.

**Comment by yorickpeterse** [comment]
Another challenge: iterators are often composed together, combined with using closures. These types may be constructed in function `A` while used in some caller `B`. The result is that escape analysis might not be able to determine whether the data escapes or not and thus has to conservatively assume that it _does_ escape. The result is that this combination likely results in many cases where we can't stack allocate the data.

**Comment by yorickpeterse** [comment]
A different approach is to do the following: instead of escape analysis, we introduce the syntax/semantics `type uni T` and `fn uni T`. A type defined using `uni` exposes its borrows as `uni ref T`/`uni mut T`, the same goes for a closure. Unlike regular `uni` borrows, these borrows allow non-sendable types to be passed/returned, with the exception of `uni` borrows themselves. Or more precisely, they only enforce the "you can't assign/pass them around" restrictions.

This would be safe because it's never possible to construct a regular `mut T` or `ref T` to a `type uni T`, and `uni ref T` and `uni mut T` values can't stick around (e.g. in a local variable).

Because of these restrictions we _can_ allow field assignments, which in turn makes `type uni` types useful for e.g. iterators (which often assign fields new values), and `fn uni` closures useful for closures that you just want to call and not borrow multiple times (the latter being quite rare to begin with).

Note that `type uni` doesn't mean an instance of such a type is automatically `uni`, because it can still contain/produce borrows to data outside of it. Instead it merely means that you can't _borrow_ it the usual way. To make it `uni` such that you can e.g. send it to another process, you'd have to combine it with `recover` as usual.

In case of `std.iter.Stream`, it means its definition would be as follows:

```inko
type pub uni Stream[T] {
  let @func: fn uni -> Option[T]

  fn pub static new(func: fn uni -> Option[T]) -> Stream[T] {
    Stream(func)
  }
}
```

When calling methods on a `type uni`, the receiver is passed as a pointer for `fn` and `fn mut` methods, and by value for `fn move`.

This approach does have some downsides:

- There's some conflict between `type inline` and `type uni` in terms of when to use them. In some cases it might be obvious (e.g. you want to "borrow" the data), but in other cases it might not be obvious which one of the two you should use
- It requires the developer to actually think about whether the data is going to be borrowed the usual way or not, i.e. it complicates the development experience
- It complicates the type checker, which is already rather complicated as-is

The first downside is the biggest one, as I really don't like it when there are multiple competing ways of doing (more or less) the same thing, because it results in a mess. For example, Python famously claims "there should be only one way to do things", only to end up with urllib/urllib2/urllib3 and similar situations.

On the other hand, the benefit of this approach is that the resulting behavior is consistent/predictable: a `type uni` type _always_ goes on the stack, regardless of how it's passed around. This can result in more predictable performance when used where necessary.

**Comment by yorickpeterse** [comment]
`type uni` types would come with a potentially annoying restriction: methods defined on such a type would have `self` exposed as `uni ref T` or `uni mut T`. Values of such types can't be captured by closures, as doing so would violate the aliasing rules/restrictions.

We also run into an issue similar to the one described in https://github.com/inko-lang/inko/issues/787: trait default methods may borrow `self`, but that wouldn't be sound for `type uni` types.

I think that in order to solve both, we essentially have to treat `self` as `uni ref T`/`uni mut T` as proposed above (= can return anything, just can't store it). That wouldn't be enough to account for expressions such as `self.something()` where `something()` returns a borrow of `self`, so we'd need something extra for that.

**Comment by yorickpeterse** [comment]
Skimming through some recent papers related to escape analysis ([link](https://scholar.google.com/scholar?start=10&q=%22escape+analysis%22&hl=nl&as_sdt=0,5&as_ylo=2010&as_vis=1)), it seems that the usual stack allocation percentages hovers around 10% on average, with some cases of 20-30%. IIRC this matches findings from papers published in the late 90s/early 2000s. If this holds true across languages (or more specifically, Inko), then I'm not sure a 10% reduction in heap allocations is worth the compile-time complexity.

**Comment by yorickpeterse** [comment]
I think that for closures we can just make `fn uni` the default behavior. I honestly can't think of a scenario where you'd want multiple mutable borrows of a closure (something that isn't even possible in Rust), so producing `uni mut fn ...` when borrowing a closure should be sufficient. 

What we _should_ do is change the capturing logic such that one closure capturing another closure captures it by _value_, as this new setup wouldn't allow capturing by borrow. This really should be the default anyway as capturing closures by borrow is almost always going to result in them being dropped prematurely.

**Comment by yorickpeterse** [comment]
My current thinking is to introduce `static ref T` and `static mut T` borrows. These can't be passed around similar to `uni ref T` and `uni mut T` borrows, but they don't impose the sendability restrictions. Similar to the unique borrows, there's no actual syntax for such borrows such that you can never specify such a type directly.

**Comment by yorickpeterse** [comment]
There's in fact an area in the standard library where we mutably borrow a closure: `std.array.stable_sort` is a recursive method and borrows a closure used for the comparison. We'd have to change this to return the closure when it's done such that it can be reused, or find a different approach (e.g. an iterative merge sort).

**Comment by yorickpeterse** [comment]
Per https://github.com/inko-lang/inko/issues/786#issuecomment-2611415807 we'll look into stack allocating closures as part of 0.19.0.

**Comment by yorickpeterse** [comment]
I quickly hacked together a method that checks how many allocations are dropped in the same scope, taking into account any moves of the original register containing the allocation. For the standard library there are 5407 heap allocation sites, but only 150 (2.77%) of those are dropped in the same method. If I increase the inline weight from 100 to 200 this only goes up to 2.86%. If I increase the threshold to 500, we're at 7.46%.

In other words, we'd have to inline much more aggressively for this to have any impact that's worth the trouble.

**Comment by yorickpeterse** [comment]
[OpenFlow](https://github.com/yorickpeterse/openflow) seems to benefit a bit more from this approach: if we only count closure allocations, then without the inlining hack mentioned above then about 16% of the closure allocations can go on the stack. Enabling the closure inlining hack only bumps that up to 19%.

For [clogs](https://github.com/yorickpeterse/clogs) the results are very similar.

**Comment by yorickpeterse** [comment]
Another way of approaching this is the following: the allocation/drop check hack mentioned is _very_ basic and likely a lot less accurate compared to escape analysis. If such a basic hack _can_ reduce allocations by (in the best case) around 20%, then perhaps escape analysis _can_ do a much better job compared to the usual 8-10% mentioned by the various papers linked in the issue description. For example, if we can stack allocate say 30% of all allocations that's still pretty good, provided the escape analysis implementation isn't overly complex.

**Comment by yorickpeterse** [comment]
Working on https://github.com/inko-lang/inko/pull/798 I realized the following: since we have single ownership and move semantics, determining if values escape should be easier than thought: given some value V, if V is unconditionally _dropped_ at every return then we know for a fact it doesn't escape the call. If V is merely _moved_ or dropped conditionally, we have to assume that it might escape. Using this information we can then annotate arguments as "escape" or "doesn't escape" and use that for inter-procedural escape analysis.

For dynamic dispatch we could produce this information for all potential targets, then "union" the results together: if given an argument A and for one or more potential targets A might escape, we treat it as escaping for all targets. Only if it doesn't escape for _any_ of the targets would we consider it as non-escaping.

With that said, we'd likely still be looking at _maybe_ 10-20% fewer allocations in the best case. That's not _bad_, but I'm not sure it's worth it either.

**Comment by yorickpeterse** [comment]
I looked into this again and implemented another basic PoC that essentially looks for allocations with corresponding `free` calls in the same method, and if so flags them as non-escaping. Similar to past experiments and the various scientific papers on the subject, this produces a "promote heap to stack" percentage of roughly 15% for projects such as https://github.com/yorickpeterse/shost and https://github.com/yorickpeterse/openflow.

While we can likely increase that number by performing analysis across methods, I strongly doubt that will increase is dramatically to e.g. 50% or more. Even with the perfect analysis, a simple `return` may flag a value as escaping. Similarly, storing data in pointers would cause values to be marked as escaping, meaning that storing values in e.g. an `Array` would result in them escaping and thus being heap allocated, even if the `Array` itself doesn't escape the surrounding scope.

On the other hand, if we allow `mut` methods for `inline` types and pass the receiver by pointer/reference instead of by value (just like Swift) then we can use `inline` types for e.g. iterators and likely cut down the amount of heap allocations dramatically. This was initially (and still not) allowed because of it potentially being confusing as field re-assignments of `inline` types won't affect borrows, but with this restriction being lifted for `move` methods already it may be time to do the same for `mut` methods.

**Comment by yorickpeterse** [comment]
I have a PoC of allowing `inline` values to be updated in-place. This however reveals a few issues:

1. We essentially have to lie about the type of `self` in `fn` and `fn mut` methods for `inline` types: they are `ref T` and `mut T` respectively, but internally become a `Pointer[T]` so we can tell the difference between "self as a pointer" and an explicit borrow of self. This requires a bunch of changes in the type specialization stage to make this work
2. Reading fields still results in a copy of the stack data rather than using the pointer as-is. This means updating `inline` values in fields currently doesn't work. This working is crucial for iterators as these make heavy use of closures and capturing, which translates to the values being stored in fields.
3. Allowing in-place updates of `inline` values in fields can only be safely done if the field isn't `mut`, otherwise a new value may be assigned during the update and invalidate the pointee of the in-place update. Swift guards against this with a runtime check, but I don't want to implement this for Inko
4. At the same time, re-assigning fields during an in-place update is quite rare so not allowing in-place updates for `mut` fields is annoying
5. For `copy` types we have to keep passing data around by value and not allow in-place updates, as types such as e.g. `Int` and `Float` expect to have the same representation and semantics of value types and not references
6. Reading from e.g. an array would still produce a new stack copy and thus won't effectively doesn't allow in-place updates of the same original data

**Comment by yorickpeterse** [comment]
I spent some more time experimenting with escape analysis. Some notes:

1. We don't forcefully inline `$dropper` calls. This means that when called on a stack allocated value, the `free()` call inside `$dropper()` tries to free non-heap memory. To resolve that we need to either _always_ inline `$dropper` so we can eliminate the `free()`, or include a runtime check to only call `free()` when operating on a heap value
2. For closures we use the `call_dropper` instruction as we don't specialize over closures. We'd have to devirtualize closure calls (#802) and drops, otherwise many closures won't be stack allocated
3. Based on my current PoC it seems we _should_ be able to stack allocate more than 10-20% of heap values even without whole-program analysis, though this ignores the fact that my current PoC segfaults in some cases (e.g. for a simple HTTP hello world). Generally the issue seems to be calling `free` on stack allocated values
4. The most common candidates thus far seem to be `Array`, `ByteArray`, `SipHasher13` and closures

For item 1 the runtime check would involve a bit set in the object header somewhere which we then check prior to calling `free`. 

For stack allocated values the type pointer might be used to obtain the type name in the event borrows still exist when the value is dropped, so we can't replace it entirely with some sort of flag. However, we _can_ tag the lowest bit and check its presence to determine if a `free` call is necessary. This check only needs to take place for heap types, and the cost of unmasking a bit and checking it is cheap enough.

Of course ideally we don't do this at all and inline `$dropper` always. This _does_ require that we don't _also always_ inline `drop` because that could result in a code explosion. If we inline `$dropper` then _also_ inlining `drop` is redundant anyway as there's usually not much in there that would benefit from inlining.

**Comment by yorickpeterse** [comment]
After today's refactoring of my PoC I believe it's now mostly correct, or at least it doesn't crash what it crashed before. However, by virtue of being more accurate (or more precisely, by not being horribly inaccurate) it seems the percentage of heap values promoted to the stack is back to roughly 10% in most cases.

To bump that up we'd really need to perform whole-program analysis as even simple patterns such as calling a moving method (= which is now considered an escape of the receiver) can inhibit stack promotion.

Whole program analysis would essentially run in two stages:

1. Perform method-local analysis first and promote allocations where possible. For arguments, record whether they escape or not
2. Repeat the analysis but use the argument escaping statistics to handle passing owned values to owned arguments, promoting values to the stack even if they're moved as an argument as long as the callee doesn't cause the value to escape. Something similar would need to be done for moving receivers

This second stage should probably use our strongly connected components traversal logic (currently specific to the inliner) so we can process leaf nodes first then work our way up.

Now the question is to what degree this is going to actually matter vs changing certain APIs, and if the complexity is worth it. For example, hashing would benefit as temporary `SipHasher13` instances could be promoted to the stack easily (without global analysis this depends on the `finish` method being inlined or not, since it's a moving method). However, one could achieve similar results by changing the hashing API in such a way that `SipHasher13` could be made an `inline` or `copy` type.

**Comment by yorickpeterse** [comment]
I've got whole-program/inter-procedural escape analysis up and running. Currently it doesn't account for cases where a stack allocated value is passed to a non-inlined method and then free'd, causing a crash. There also seem to be some other issues with incorrect code being generated. We also don't yet handle recursive calls, though I'm not sure this actually needs any special care when it comes to escape analysis.

In terms of the amount of allocations that can be promoted to stack allocations, this percentage now hovers around 50% for programs such as Openflow, shost, and a simple HTTP client+server example. Considering we don't devirtualize closures (and thus don't inline them) I think I should be able to further increase that number, as closures are a significant source of heap allocations.

So basically the next steps are (roughly):

1. Extend the `Free` instruction so it can transparently handle heap allocated values and heap-to-stack promoted values
2. Ensure the generated code is in fact correct and doesn't cause any crashes
3. Investigate where we can make the algorithm less conservative
4. #802

**Comment by yorickpeterse** [comment]
A nice side effect of conditional `free` calls is that it would also solve #777: by not marking constant strings as "heap allocated" we'd just skip the `free` call should the reference count ever overflow and wrap around to zero.
